R/locale.R

```diff
@@ -1,84 +1,26 @@
-overrideWindowsLocale <- function(locale) {
-  map <- list()
-  map[["el_EL"]] <- "el_GR"
-  if (locale %in% names(map)) {
-    locale <- map[[locale]]
-  }
-  return(locale)
-}
-
 detectLocale <- function() {
-  sysName <- Sys.info()[["sysname"]]
-  if (identical(sysName, "Windows")) {
-    locale <- detectLocale.Windows()
-  } else {
-    locale <- detectLocale.Unix()
-  }
-  return(locale)
-}
-
-detectLocale.Unix <- function() {
-  unlist(strsplit(Sys.getlocale("LC_CTYPE"), ".", fixed = TRUE))[1]
-}
-
-detectLocale.Windows <- function(useCache =
-                                  getOption("rsconnect.locale.cache", TRUE)) {
-
-  # default locale
-  locale <- "en_US"
-
-  cacheFile <- localeCacheFile()
-  if (file.exists(cacheFile) && useCache) {
-
-    # get chached
-    cache <- as.list(readDcf(cacheFile, all = TRUE))
-
-    locale <- unlist(cache$locale)
-
-  } else {
-
-    tryCatch({
-
-      # get system locale
-      locale <- systemLocale()
-
-      # write the user info
-      write.dcf(list(locale = locale),
-                cacheFile,
-                width = 100)
-
-    }, error = function(e) {
-      warning(paste0("Error detecting locale: ", e,
-                     " (Using default: ", locale, ")"), call. = FALSE)
-    })
-  }
-  return(overrideWindowsLocale(locale))
-}
-
-localeCacheFile <- function() {
-  normalizePath(file.path(rsconnectConfigDir(), "locale.dcf"), mustWork = FALSE)
-}
-
-systemLocale <- function() {
-  message("Detecting system locale ... ", appendLF = FALSE)
-
-  # get system locale
-  info <- systemInfo()
-  raw <- as.character(info[[20]])
-  parts <- strsplit(unlist(strsplit(raw, ";",  fixed = TRUE)), "-", fixed = TRUE)
-
-  if (length(parts[[1]]) >= 2) {
-    # normalize locale to something like en_US
-    locale <- paste(tolower(parts[[1]][1]), toupper(parts[[1]][2]), sep = "_")
+  if (!isWindows()) {
+    locales <- strsplit(Sys.getlocale("LC_CTYPE"), ".", fixed = TRUE)[[1]]
+    locales[[1]]
   } else {
-    locale <- paste(tolower(parts[[1]][1]), toupper(parts[[1]][1]), sep = "_")
+    locale <- HCU_registry_key("Control Panel\\International\\User Profile")$Languages[[1]]
```

Comment:
Actually, it is better to only use the other registry key, because this one can return the language code only, without the region. E.g. for Hungarian I get
```r
> HCU_registry_key("Control Panel\\International\\User Profile")$Languages
[1] "hu"    "en-US" "de-DE" "es-ES"
> HCU_registry_key("Control Panel\\International")$LocaleName
[1] "hu-HU"
```

OTOH, the second one is probably just the duplication of the language code. Which ends up being a correct locale for Hungary: `hu-HU`, but the same thing for Greek and `el` does not work. 

The PDF here has the list of the valid locale codes: https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-lcid/70feba9f-294e-491e-b6eb-56532684c37f

Reply:
Another good resource, where they suggest the `$LocaleName` registry entry: https://ss64.com/locale.html

Reply:
Does that linked doc imply that sometimes the `LocaleName` might include a script identify that we should strip?

Reply:
Good point, the table at https://ss64.com/locale.html is auto-generated, and has locales with script ids, even without a region sometimes, e.g.:
```
az-Cyrl
az-Cyrl-AZ
```

So I _think_ we need to split by `-`, keep the first part, and remove the part that is not uppercase (if any). If we always need a region code, then add one if needed. 

There is potentially also a sort order at the end, separated by an underscore, e.g. `hu-HU_technl` but I am not sure if `LocaleName` would ever include that.